contentpal/src/test/java/org/dmfs/android/contentpal/operations/CountedTest.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.android.contentpal.operations;
+
+import android.content.ContentProviderOperation;
+
+import org.dmfs.android.contentpal.Operation;
+import org.dmfs.android.contentpal.SoftRowReference;
+import org.dmfs.android.contentpal.TransactionContext;
+import org.dmfs.optional.Optional;
+import org.junit.Test;
+
+import static org.hamcrest.Matchers.sameInstance;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+
+/**
+ * @author Marten Gajda
+ */
+public class CountedTest
+{
+    @Test
+    public void testReference() throws Exception
+    {
+        final Optional<SoftRowReference<Object>> testReference = mock(Optional.class, new FailAnswer());
+        Operation<Object> mockOperation = mock(Operation.class, new FailAnswer());
+        doReturn(testReference).when(mockOperation).reference();
+
+        assertThat(new Counted<>(1, mockOperation).reference(), sameInstance(testReference));
+    }
+
+
+    @Test
+    public void testContentOperationBuilder() throws Exception
+    {
+        final ContentProviderOperation.Builder testBuilder = mock(ContentProviderOperation.Builder.class, new FailAnswer());
```

Comment:
These tests looks very good now! I didn't know this `Answer` stubbing feature, nice.
Much easier for me to get a quicker understanding of the tests now.

One totally optional thing maybe, to not re-use that `testBuilder` here but use a separate instance, that would make it even clearer for me. Like:
```
        ContentProviderOperation.Builder mockBuilder = mock(ContentProviderOperation.Builder.class, new FailAnswer());
        ContentProviderOperation.Builder resultBuilderDummy = mock(ContentProviderOperation.Builder.class, new FailAnswer());
        doReturn(resultBuilderDummy).when(mockBuilder).withExpectedCount(10);
```

Also as I see these mocking tests with Mockito can use differrent kind of 'test doubles'. I am wondering, would it make sense to start to try to use a terminology for naming the variables, based on this categorization for example?:
https://martinfowler.com/articles/mocksArentStubs.html

> Dummy objects are passed around but never actually used. Usually they are just used to fill parameter lists.
> Fake objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an in memory database is a good example).
> Stubs provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.
> Spies are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.
> Mocks are what we are talking about here: objects pre-programmed with expectations which form a specification of the calls they are expected to receive.

Probably it wouldn't always be clear, but still for example a distinction between a Mock and a Dummy - which can both be created technically as Mockito.mock() - may make things clearer in test. Or a mock() with `FailAnswer` could be viewed as a stub if it is not set up for other expectation, if it is then it becomes a Mock.
Not important of course, just a general idea/question.

Reply:
I've deliberately chosen `doReturn(testBuilder).when(testBuilder).withExpectedCount(10);` because it models the reality in which [`withExpectedCount(10)` always returns the same instance](https://developer.android.com/reference/android/content/ContentProviderOperation.Builder.html#withExpectedCount(int))

Good Point about the names. This can make a lot of sense for more complex tests. For small tests, although still a good idea, the benefit is probably not so big.

Reply:
Oh, I see the builder returns itself.. I understand it then.